### PR TITLE
[To rel/0.13][IOTDB-3843] Refine the using of setting read-only

### PR DESCRIPTION
**Pull request by HeimingZ**

### CODE REVIEW

File: server/src/assembly/resources/conf/iotdb-engine.properties [hunk 1]
```diff
@@ -193,6 +193,10 @@ timestamp_precision=ms
 # Datatype: long
 # default_ttl=36000000
 
+# Shutdown system or set it to read-only mode when unrecoverable error occurs.
+# Datatype: bool
+# enable_read_only = true
```
Comment (HTHou) [inline server/src/assembly/resources/conf/iotdb-engine.properties:198]:
```suggestion
# enable_read_only=true
```

File: server/src/main/java/org/apache/iotdb/db/conf/IoTDBConfig.java [hunk 2]
```diff
@@ -1384,12 +1388,39 @@ public void setSessionTimeoutThreshold(int sessionTimeoutThreshold) {
...
+      if (enableReadOnly) {
+        logger.error(
+            "System mode will be set to {}! Only query statements are permitted!",
+            newStatus,
+            new RuntimeException("System mode is set to read-only"));
+      } else {
+        logger.error(
+            "Unrecoverable error occurs! Shutdown system directly when read-only mode is disabled.",
+            new RuntimeException("System mode is set to read-only"));
+        System.exit(-1);
```
Comment (qiaojialin) [inline server/src/main/java/org/apache/iotdb/db/conf/IoTDBConfig.java:1418]: Do not shutdown system when disk is full.

File: server/src/main/java/org/apache/iotdb/db/conf/IoTDBConfig.java [hunk 3]
```diff
@@ -1384,12 +1388,39 @@ public void setSessionTimeoutThreshold(int sessionTimeoutThreshold) {
...
-    this.readOnly = readOnly;
+  public SystemStatus getSystemStatus() {
+    return status;
+  }
+
+  public void setSystemStatus(SystemStatus newStatus) {
+    if (newStatus == SystemStatus.READONLY) {
+      if (enableReadOnly) {
+        logger.error(
+            "System mode will be set to {}! Only query statements are permitted!",
```
Comment (qiaojialin) [inline server/src/main/java/org/apache/iotdb/db/conf/IoTDBConfig.java:1411]: Keep the log consistent with previous:

change system mode to read-only

File: server/src/main/java/org/apache/iotdb/db/conf/SystemStatus.java [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iotdb.db.conf;
+
+/** Status of current system */
+public enum SystemStatus {
+  /** System can read and write normally */
+  NORMAL,
+  /** Only read statements are permitted */
```
Comment (qiaojialin) [inline server/src/main/java/org/apache/iotdb/db/conf/SystemStatus.java:25]:
```suggestion
  /** Only query statements are permitted */
```

File: server/src/main/java/org/apache/iotdb/db/conf/SystemStatus.java [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+  READONLY,
```
Comment (qiaojialin) [inline server/src/main/java/org/apache/iotdb/db/conf/SystemStatus.java:26]: Only directly set read-only when disk is full.

Add a new status: ERROR.
And reset to read-only or shutdown system when set system to error status.

File: server/src/main/java/org/apache/iotdb/db/engine/storagegroup/TsFileProcessor.java [hunk 6]
```diff
@@ -1148,11 +1151,11 @@ public void flushOneMemTable() {
         }
       } catch (Exception e) {
         logger.error(
-            "{} meet error when flush FileMetadata to {}, change system mode to read-only",
+            "{}: {} marking or ending file meet error",
             storageGroupName,
-            tsFileResource.getTsFile().getAbsolutePath(),
+            tsFileResource.getTsFile().getName(),
```
Comment (qiaojialin) [inline server/src/main/java/org/apache/iotdb/db/engine/storagegroup/TsFileProcessor.java:1156]:
```suggestion
            tsFileResource.getTsFile().getAbsolutePath(),
```
Use absolute path is better for debugging

File: server/src/main/java/org/apache/iotdb/db/engine/storagegroup/TsFileProcessor.java [hunk 7]
```diff
@@ -1148,11 +1151,11 @@ public void flushOneMemTable() {
...
             e);
-        IoTDBDescriptor.getInstance().getConfig().setReadOnly(true);
+        // truncate broken data
```
Comment (qiaojialin) [inline server/src/main/java/org/apache/iotdb/db/engine/storagegroup/TsFileProcessor.java:1158]:
```suggestion
        // truncate broken metadata
```

File: server/src/main/java/org/apache/iotdb/db/engine/storagegroup/VirtualStorageGroupProcessor.java [hunk 8]
```diff
@@ -1192,24 +1193,34 @@ public void submitAFlushTaskWhenShouldFlush(TsFileProcessor tsFileProcessor) {
...
+        IoTDBDescriptor.getInstance().getConfig().setSystemStatus(SystemStatus.READONLY);
+        break;
+      } catch (IOException e) {
+        if (retryCnt < 3) {
+          logger.warn("meet IOException when creating TsFileProcessor, retry it again", e);
+          retryCnt++;
+        } else {
+          logger.error(
+              "meet IOException when creating TsFileProcessor, change system mode to read-only", e);
+          IoTDBDescriptor.getInstance().getConfig().setSystemStatus(SystemStatus.READONLY);
```
Comment (qiaojialin) [inline server/src/main/java/org/apache/iotdb/db/engine/storagegroup/VirtualStorageGroupProcessor.java:1219]:
```suggestion
          IoTDBDescriptor.getInstance().getConfig().setSystemStatus(SystemStatus.ERROR);
```

File: session/src/main/java/org/apache/iotdb/session/Session.java [hunk 9]
```diff
@@ -421,6 +422,10 @@ public SessionConnection constructSessionConnection(
     return new SessionConnection(session, endpoint, zoneId);
   }
 
+  public SystemStatus getSystemStatus() throws IoTDBConnectionException {
+    return defaultSessionConnection.getSystemStatus();
+  }
```
Comment (HTHou) [inline session/src/main/java/org/apache/iotdb/session/Session.java:427]: Add this method in sessionPool?

File: server/src/assembly/resources/conf/iotdb-engine.properties (hunk 1 above)
Comment (HeimingZ) [reply server/src/assembly/resources/conf/iotdb-engine.properties]: Rename it to read_only_strategy.

File: server/src/main/java/org/apache/iotdb/db/conf/IoTDBConfig.java (hunk 2 above)
Comment (HeimingZ) [reply server/src/main/java/org/apache/iotdb/db/conf/IoTDBConfig.java:1417]: Add ReadOnlyStrategy enumeration and read_only_strategy parameter.

File: server/src/main/java/org/apache/iotdb/db/conf/SystemStatus.java (hunk 4 above)
Comment (HeimingZ) [reply server/src/main/java/org/apache/iotdb/db/conf/SystemStatus.java:25]: Fixed.

File: server/src/main/java/org/apache/iotdb/db/engine/storagegroup/TsFileProcessor.java (hunk 6 above)
Comment (HeimingZ) [reply server/src/main/java/org/apache/iotdb/db/engine/storagegroup/TsFileProcessor.java]: Fixed.

File: server/src/main/java/org/apache/iotdb/db/engine/storagegroup/TsFileProcessor.java (hunk 7 above)
Comment (HeimingZ) [reply server/src/main/java/org/apache/iotdb/db/engine/storagegroup/TsFileProcessor.java]: Fixed.

File: session/src/main/java/org/apache/iotdb/session/Session.java (hunk 9 above)
Comment (HeimingZ) [reply session/src/main/java/org/apache/iotdb/session/Session.java:427]: Fixed.

**Review by qiaojialin — APPROVED** [review]

**Pull request merged**